capellambse/model/diagram.py

```diff
@@ -302,6 +302,36 @@ def type(self) -> modeltypes.DiagramType:
             LOGGER.warning("Unknown diagram type %r", sc)
             return modeltypes.DiagramType.UNKNOWN
 
+    @property
+    def activated_filters(self) -> list[str]:
+        """Return a sequence of activated filters"""
+        return aird.parser._filters.get_filters_from_diagram(
+            self._model._loader, self._element.uid
+        )
+
+    def add_activated_filter(self, fltr_name: str) -> None:
+        """Add an activated filter to the diagram.
+
+        Writes a new <activatedFilters> XML element to the
+        <diagram:DSemanticDiagram> XML element. If the `fltr_name` is
+        not apparent in :data:`aird.parser.GLOBAL_FILTERS` as a key
+        it can not be applied when rendering. It should still be visible
+        in the GUI.
+        """
+        aird.parser._filters.add_filter_to_diagram(
+            self._model._loader, self._element, fltr_name
+        )
+        self._create_diagram()
```

Comment:
Note that `self._create_diagram()` *returns* the rendered diagram, however the return value is not used in this method. Therefore this call is superfluous. I assume what you wanted to do here was to refresh the render cache.

Instead, we should merely *invalidate* it so that a full re-render will happen on the next call to `render()`. Otherwise multiple calls to `add_activated_filter` would cause multiple intermediate renders to happen (one for each added filter), which is a significant waste of resources. And on top of that, there's the possible use case of programmatically adding the filters on several diagrams at once and not rendering them with capellambse at all, but saving the filters for further use in Capella. (Adding a public method on `AbstractDiagram` to `invalidate_cache()` seems like an overdue change anyways, so let's do that while we're at it.)

I believe this should also address your comment from the opening post:

> Make sure to add/remove filters **before** rendering diagrams such that the `capellambse.aird.parser._filters.applyfilters` is called.

Because, after invalidating the cache, this function would in fact be called again during the next `render()`.

Reply:
We did it a bit different. We should have a talk/call about this.